### ISSUE TITLE
distributed training?

### ISSUE THREAD

**Issue by JiayiLy** [koyeongmin/PINet_new#16]
It seems that your code doesn't support distributed training. I’m trying to translate your code to mmdetection? I have some doubts:
1. How much does the hard sampling improve the final performance?
2. The scheduler in your code is quite strange. Can I use another one such as CosineAnnealing or MultiStep?

Looking forward to your reply. Thanks!

**Comment by koyeongmin** [comment]
1. Hard sampling module is not a major factor, but in CuLane dataset, the hard sampling shows more improvement than TuSimple dataset.
2. Yes!, My scheduler is not good.. You can use another scheduler!
Thank you!

**Comment by JiayiLy** [comment]
It seems that learning rate will never change during training in your code? I have tried a CosineAnnealingLR, but the peformance degrades a lot. 
![Screenshot from 2020-10-30 15-30-07](https://user-images.githubusercontent.com/16137282/97672516-2796b600-1ac5-11eb-8858-851a2a2da578.png)

The confidence loss and attention loss look strange. Do you have any idea about this? Thank you

**Comment by JiayiLy** [comment]
Why do you use a regression loss (L2 loss) when computing confidence loss? Have you ever tried a classification loss?

**Comment by JiayiLy** [comment]
Can you show your result on tusimple dataset without hard sampling? I got a result so follows:

> Accuracy 0.9388447417068856
FP 0.09802043750641848
FN 0.062065660196501254

**Comment by koyeongmin** [comment]
https://github.com/koyeongmin/PINet
This is the previous version of PINet, and this model got 96.64% (accuracy) without hard sampling.
Unfortunately, the result of the current version without hard sampling does not remain, but I remember the performance is similar to the previous version.
Thank you!

**Comment by JiayiLy** [comment]
Thank you so much. I will try to tune some other parameters. By the way, do you remember the result on tusimple dataset without data agumentation?

**Comment by koyeongmin** [comment]
Sorry, I don't remember it because I have always used data augmentation after first some trials.

**Comment by JiayiLy** [comment]
OK. Do you have any idea about the existence loss. I see you comment out 
``confidence` = torch.sigmoid(confidence)`
I tried sigmoid on the confidence, the loss curve is perfect but the accuracy drops a bit.

**Comment by koyeongmin** [comment]
I also tried other loss like sigmoid and cross-entropy, but performance is not good...

**Comment by jiayily** [comment]
Thank you so much for your patience and your nice work.

**Comment by jiayily** [comment]
@koyeongmin Have you ever tried dice loss and lovasz hinge loss? I tried these losses on the confidence branch with thresh_point=0.5. The performance got a little improvement. I think the dice loss is suited for tusimple evaluation and the lovasz hinge loss is suited for culan evaluation (miou metric).

**Comment by koyeongmin** [comment]
Thank you! I have not tried these losses. I wiil try at future works. Thank you!

**Comment by jiayily** [comment]
Hello, have you ever studied self-attention-distillation (https://github.com/cardwing/Codes-for-Lane-Detection)? I think your attention loss have the same intuition with it. But how does it work when we only have one hourglass?

**Comment by mvish7** [comment]
@Jiayi719 did you succeed in distributed training of PINet?? and did you use PyTroch DDP for it??

**Comment by jiayily** [comment]
@mvish7 Yes. I used DistributedDataParallel to implement it.

**Comment by Wolfwjs** [comment]
Hello，could you please share your code or core code of PINet based mmdetection? thx!

**Comment by mvish7** [comment]
Hi @Wolfwjs 
Unfortunately I cant share the code. 
I used torch DDP. The problem I had was related to not using all the outputs of the model. I.e. i used some detach commands in wrong places.

**Comment by Wolfwjs** [comment]
thank you for your reply！ @mvish7

**Comment by Wolfwjs** [comment]
Hello，have you successfully run PINet based on mmdection? @Jiayi719
